### ISSUE TITLE
Pin maven version on GitHub action

### ISSUE THREAD

**Issue by philsmart** [ukf/ukf-members#11]
actions/setup-java uses the latest Maven version. It does not appear to have a setting to pin this to a specific version (they have an issue about this https://github.com/actions/setup-java/issues/40). In reality, they say it uses whichever version is installed on the runner (not sure how to control that).

There are other 'setup-maven' type actions on the marketplace which may allow this. Although I am a bit cautious about using a random user's GitHub action over a more standard pipeline using the official GitHub actions.

We should look to try and control this perhaps. Needs investigation.

**Comment by iay** [comment]
This sounds like it would be important, and I would be a little surprised if there wasn't a way to do it.

On the other hand, if there _isn't_ then I think it's a limitation that we can accept, rather than (e.g.) moving heaven and earth to install a specific, pinned, version of Maven. Most of the variability comes from the plugins, and we pin all the ones we think are significant in the POM (and parent POMs). So changes in the version of the Maven _engine_ shouldn't affect us, and if they do we probably have something that should be corrected. Maven has historically been fairly good about pointing things like this out.

So yes, you should look into this but if it turns out to be anything but trivial, I'm happy to just close this without action.

**Comment by philsmart** [comment]
Well, I can not disagree with my former self. 

I think that to proceed, we would need to do one of the following:
 1. Add some setup commands to pull in a specific version of Maven into the Ubuntu runner, or 
 2. Use a random Maven action from the Marketplace (for example: https://github.com/marketplace/actions/setup-maven-action).
 3. Add a Maven wrapper to the project with a pinned version.

Of those, I’m not sure how robust option (1) would be in the long term, or how trustworthy option (2) is. So I guess we could try adding a Maven wrapper, which we would need to commit to the repository—although I am not sure how clean that is; I am not sure if it just adds complexity.

**Comment by philsmart** [comment]
Ah, right, we actually use the maven-settings-action from the s4u folk in ukf-mda:
```
name: Set up Maven
      uses: s4u/maven-settings-action@v3.0.0
      with:
```
I guess my mindset is more concerned with supply chain issues these days, but if we are already using it...

**Comment by philsmart** [comment]
Well, it does not look like you can use `s4u/maven-settings-action` to set the maven version either, it just adds things to the settings file.

**Comment by philsmart** [comment]
There is the setup version action (as I listed in the options above): https://github.com/marketplace/actions/setup-maven-action. But that also handles the JDK version, which overlaps with our setup java action.

**Comment by philsmart** [comment]
Right, actually reading https://github.com/marketplace/actions/setup-maven-action, it is a composite action which replaces [actions/checkout](https://github.com/marketplace/actions/checkout), and [actions/setup-java](https://github.com/marketplace/actions/setup-java-jdk), and adds the ability to set the Maven version using the https://github.com/stCarolas/setup-maven/ plugun. So we could replace the other actions with this one, or just try out the setup-maven action inside what we already have. I'll try that in a branch.

**Comment by philsmart** [comment]
We tested the setup-maven plugin, which, while it seemed to work, I concluded:

> On reflection, I’m perhaps inclined to avoid introducing another action that we’d need to update and potentially debug. Given that, it’s probably best to close this issue without merging these changes.
